runtime/src/main/java/org/corfudb/runtime/view/AddressSpaceView.java

```diff
@@ -554,7 +604,46 @@ public void invalidateClientCache() {
             }
         }
 
-        result.forEach(this::checkLogData);
+        return checkForTrimmedAddresses(result, cacheOnClient, ignoreTrimmed);
+    }
+
+    /**
+     * Check list of returned data for trimmed data.
+     *
+     * @param result list of read data
+     * @param cacheOnClient flag whether the valid data should be cached on the client or not.
+     * @param ignoreTrimmed flag whether trimmed data leads to TrimmedExceptions or ignored.
+     *
+     * @return map with valid data (not trimmed) or TrimmedException is thrown is data is trimmed and
+     * flag is not set.
+     */
+    private Map<Long, ILogData> checkForTrimmedAddresses(Map<Long, ILogData> result,
+                                                         boolean cacheOnClient, boolean ignoreTrimmed) {
+        List<Long> trimmedAddresses = new ArrayList<>();
+
+        for (Map.Entry<Long, ILogData> entry : result.entrySet()) {
+            // Add trimmed addresses to list
+            if (!checkLogData(entry.getKey(), entry.getValue(), false)) {
+                trimmedAddresses.add(entry.getKey());
+            } else {
+                if (cacheOnClient) {
+                    // After fetching a value, we need to insert it in the cache.
+                    // Even if trimmed exceptions are thrown, we cache the valid data.
+                    result.put(entry.getKey(), cacheLoadAndGet(readCache, entry.getKey(), entry.getValue()));
```

Comment:
Same as above, I think it's not good to use cacheLoadAndGet here, read() does this.